### PR TITLE
feat: onboarding steps, navigation, & sdk logic

### PR DESCRIPTION
**Pull request by raquelmsmith**
**Don't get scared by the line count,** it's mostly SVGs and install snippets copy/pasted from ingestion - you can ignore all those 😄

## Problem

Each product will get its own onboarding, with theoretically different steps. Some steps will be similar between products (eg selecting an SDK for Product Analytics or Feature Flags) but other steps will be different or need to go in a different order.

Additionally, after this is built out, the product onboarding should be easy for other teams to work within since each team should eventually be responsible for their own onboarding flows (as they know their users best). 

So, my goal with the new onboarding developer experience is to make it work as magically as possible so teams don't have to worry about navigation logic through the onboarding, and to make it very easy to add new steps and change existing ones.

This PR lays out my current ideas for how this could all work. I'm looking for feedback from different teams:

1. Does this format feel like something you could work within to add a new onboarding step?
2. If you were to add a new product, is it fairly clear how you'd add a new onboarding for it?
3. What improvements do you see that make it easier or more robust?

## Review notes
* When looking at this PR you can **ignore anything in the sdks/logos and sdks/product-analytics folders** (where most of the line count comes from)
* I left comments in the relevant bits where you (as a product team making/updating your product onboarding flows) would have to make changes
* If you want to try this out locally:
  1. `posthog.featureFlags.override({'product-specific-onboarding':'test'})`
  2. Go to /billing and subscribe to feature flags (not product analytics!) to get your instance a license
  3. Go to `/products` and select Product Analytics (since it has SDKs added)
* Video: 

https://github.com/PostHog/posthog/assets/18598166/614191dc-8ec8-4112-a43a-6e0b80a73b3a

## Changes

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

### CODE REVIEW

File: frontend/src/scenes/onboarding/Onboarding.tsx [hunk 1]
```diff
@@ -4,141 +4,81 @@ import { useEffect } from 'react'
...
 
+const getOnboarding = (productKey: string): JSX.Element => {
+    if (productKey === 'product_analytics') {
+        return <ProductAnalyticsOnboarding />
+    } else if (productKey === 'session_replay') {
+        return <SessionReplayOnboarding />
+    } else if (productKey === 'feature_flags') {
+        return <FeatureFlagsOnboarding />
+    }
+    return <></>
```
Comment (raquelmsmith) [inline frontend/src/scenes/onboarding/Onboarding.tsx:79]: You'd have to add your product key here - probably some improvement here in terms of code quality but i think we'd need some sort of switch statement to get the correct flow

File: frontend/src/scenes/onboarding/onboardingLogic.tsx [hunk 2]
```diff
@@ -33,6 +37,36 @@ export const onboardingLogic = kea<onboardingLogicType>({
...
+                setProductKey: (_, { productKey }) => {
+                    switch (productKey) {
+                        case 'product_analytics':
+                            return urls.default()
+                        case 'session_replay':
+                            return urls.replay()
+                        case 'feature_flags':
+                            return urls.featureFlags()
+                        default:
+                            return urls.default()
```
Comment (raquelmsmith) [inline frontend/src/scenes/onboarding/onboardingLogic.tsx:65]: You'd have to define here where to go after the user completes onboarding

**Comment by tiina303** [comment]
If the user wants everything - how does the experience work?

**Comment by raquelmsmith** [comment]
> If the user wants everything - how does the experience work?

They will go through the onboarding for each product. If some steps are unnecessary (ie. they already have the snippet installed and just want to start using session replay) then the onboarding will be abbreviated, this will just act as an intro and pricing agreement of sorts.

**Comment by corywatilo** [comment]
> They will go through the onboarding for each product

How will they get back to the setup flow after installing one product? (May also be worth notating the steps you need to take if you want to enable >1.

(Another option would be to have them select all the products they want to install, _then_ begin the process – like when selecting apps to install on a smart TV – but might be overdoing it.)

Might also be worth saying everything's free / no credit card required to try?

One nit: When you hover a box, it adds the drop-shadow, but it isn't clickable. If we're going to drop shadow them on hover, we should probably link them to the same place as the button.

(FYI I was getting errors for every page beyond the `/products` page.)

image

**Comment by raquelmsmith** [comment]
> How will they get back to the setup flow after installing one product? (May also be worth notating the steps you need to take if you want to enable >1.

Current thought is that if they click on an item in the nav, if their project hasn't onboarded into that product, it'll take them to the onboarding flow. As we add more products we could maybe use the `/products` page as a listing of all available products, with types and search etc. Right now it's just a landing for immediately after signup.

> (Another option would be to have them select all the products they want to install, then begin the process – like when selecting apps to install on a smart TV – but might be overdoing it.)

That'd be cool. Will consider how to do that after I get this first iteration done!

> Might also be worth saying everything's free / no credit card required to try?

Yeah good idea, can add that to the intro page. Right now just building for the "has CC in" state, will optimize the upgrade flow once that's done.

> One nit: When you hover a box, it adds the drop-shadow, but it isn't clickable. If we're going to drop shadow them on hover, we should probably link them to the same place as the button.

Ok I wasn't sure about that but can do, thanks!

> (FYI I was getting errors for every page beyond the /products page.)

Probably because your local instance isn't licensed.... You can go to /billing and subscribe to get a license, and then it should work (as long as the product you're using isn't the one you sub'd to - still _very_ in-progress on the upgrade stuff 😅 )

**Comment by raquelmsmith** [comment]
SDK filtering works now! (I know, I know, the design and interactions need work - focusing on logic and basic flows here)

https://github.com/PostHog/posthog/assets/18598166/81dc4c6d-d59b-4b42-8711-9bc6530ef32f

File: frontend/src/scenes/onboarding/sdks/product-analytics/ProductAnalyticsSDKInstructions.tsx [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    ProductAnalyticsRNInstructions,
+    ProductAnalyticsRubyInstructions,
+} from '.'
+
+export const ProductAnalyticsSDKInstructions = {
+    [SDKKey.JS_WEB]: JSWebInstructions,
+    // add next, getsby, and others here
+    [SDKKey.IOS]: ProductAnalyticsIOSInstructions,
+    [SDKKey.REACT_NATIVE]: ProductAnalyticsRNInstructions,
+    [SDKKey.ANDROID]: ProductAnalyticsAndroidInstructions,
```
Comment (raquelmsmith) [inline frontend/src/scenes/onboarding/sdks/product-analytics/ProductAnalyticsSDKInstructions.tsx:22]: This is how you'd add support for a certain SDK to a product.

File: frontend/src/scenes/onboarding/sdks/sdksLogic.tsx [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+import type { sdksLogicType } from './sdksLogicType'
+import { ProductKey, SDK } from '~/types'
+import { onboardingLogic } from '../onboardingLogic'
+import { ProductAnalyticsSDKInstructions } from './product-analytics/ProductAnalyticsSDKInstructions'
+import { allSDKs } from './allSDKs'
+import { LemonSelectOptions } from 'lib/lemon-ui/LemonSelect/LemonSelect'
+
+export const productAvailableSDKs = {
+    [ProductKey.PRODUCT_ANALYTICS]: ProductAnalyticsSDKInstructions,
+}
```
Comment (raquelmsmith) [inline frontend/src/scenes/onboarding/sdks/sdksLogic.tsx:12]: Have to add your products SDKs here too

File: frontend/src/scenes/onboarding/sdks/sdksLogic.tsx [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+                })
+                .filter((sdk) => Object.keys(productAvailableSDKs[values.productKey || '']).includes(sdk.key))
+            actions.setSDKs(filteredSDks)
+            actions.setSourceOptions(getSourceOptions(values.productKey || ''))
+        },
+        setSourceFilter: () => {
+            actions.filterSDKs()
+            actions.setSelectedSDK(null)
+        },
+        [onboardingLogic.actionTypes.setProductKey]: () => {
```
Comment (raquelmsmith) [inline frontend/src/scenes/onboarding/sdks/sdksLogic.tsx:89]: @mariusandra this does not seem to work, do you see what might be wrong?

**Pull request merged**